processing/src/main/java/io/druid/query/aggregation/first/StringFirstBufferAggregator.java

```diff
@@ -0,0 +1,151 @@
+/*
+ * Licensed to Metamarkets Group Inc. (Metamarkets) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. Metamarkets licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.query.aggregation.first;
+
+import io.druid.java.util.common.StringUtils;
+import io.druid.query.aggregation.BufferAggregator;
+import io.druid.query.aggregation.SerializablePairLongString;
+import io.druid.query.monomorphicprocessing.RuntimeShapeInspector;
+import io.druid.segment.BaseLongColumnValueSelector;
+import io.druid.segment.BaseObjectColumnValueSelector;
+
+import java.nio.ByteBuffer;
+
+// TODO: Unit Test
+public class StringFirstBufferAggregator implements BufferAggregator
+{
+  private final BaseLongColumnValueSelector timeSelector;
+  private final BaseObjectColumnValueSelector valueSelector;
+  private final int maxStringBytes;
+
+  public StringFirstBufferAggregator(
+      BaseLongColumnValueSelector timeSelector,
+      BaseObjectColumnValueSelector valueSelector,
+      int maxStringBytes
+  )
+  {
+    this.timeSelector = timeSelector;
+    this.valueSelector = valueSelector;
+    this.maxStringBytes = maxStringBytes;
+  }
+
+  @Override
+  public void init(ByteBuffer buf, int position)
+  {
+    buf.putLong(position, Long.MAX_VALUE);
+    buf.putInt(position + Long.BYTES, 0);
+  }
+
+  @Override
+  public void aggregate(ByteBuffer buf, int position)
+  {
+    ByteBuffer mutationBuffer = buf.duplicate();
+    mutationBuffer.position(position);
+
+    Object value = valueSelector.getObject();
+
+    long time = Long.MAX_VALUE;
+    String firstString = null;
+
+    if (value instanceof SerializablePairLongString) {
+      SerializablePairLongString serializablePair = (SerializablePairLongString) value;
+      time = serializablePair.lhs;
+      firstString = serializablePair.rhs;
+    } else if (value instanceof String) {
+      time = timeSelector.getLong();
+      firstString = (String) value;
+    } else if (value != null) {
+      throw new IllegalStateException(
+          "Try to aggregate unsuported class type ["
+          + value.getClass().getName() +
+          "]. Supported class types: String or SerializablePairLongString"
+      );
+    }
+
+    long lastTime = mutationBuffer.getLong(position);
+    if (firstString != null && time < lastTime) {
```

Comment:
This looks a bug. The byteBuffer will be updated only when `firstString` is not null.

Reply:
If I understand good the logic of byte buffer aggregator, I think that it is okay. If the `firstString` value is `null`, the aggregator should skip this value and keep the not `null` value. What do you think?

Reply:
@jihoonson Maybe we can add an aggregation option `filterNullValues` (by default: `false`). What do you think?

Reply:
Do you mean adding `filterNullValues` to `aggregatorFactory`? I think it's not necessary. We already support filter to get non-null values like this:

```
  "filter" : {
    "type" : "not",
    "field" : {
      "type" : "selector",
      "dimension" : "dim",
      "value" : "" (should be null after https://github.com/apache/incubator-druid/issues/4349)
    }
  },
```

With this filter, Druid's storage layer returns only non-null values, so aggregators never read null values. This is a more consistent way to handle nulls.

Reply:
Yeah .. but this is different because the filters run when queries are performed on the aggregated data.

For example:

* 1 minute queryGranularity
* dimensions: "client"
* Metrics: stringLast("location")

And I have this example dataset

```
minute: 1, client: AAA, location: zoneA
minute: 1.1, client: BBB, location: zoneA
minute: 1.3, client: AAA, location: zoneC
minute 2, client AAA, location: zoneB
```

The aggregated rows are:

```
minute: 1, client: AAA, location: zoneC
minute: 1, client: BBB, location: zoneA
minute: 2, client: AAA, location: zoneB
```

Now, if we add some event with `client:AAA` and `location: null` at `minute: 1.4`, the output is:

```
minute: 1, client: AAA, location: null
minute: 1, client: BBB, location: zoneA
minute: 2, client: AAA, location: zoneB
```

In this case, you can't use filters. If you use filters you discard the row `minute: 1, client: AAA, location: null
`, but you can't recover the previous value `minute: 1, client: AAA, location: zoneC`. This is the motivation to add an option to the aggregatorFactory, with the option you can allow enable/disable discard null values at indexing time.

What do you think?

Reply:
> In this case, you can't use filters. If you use filters you discard the row minute: 1, client: AAA, location: null, but you can't recover the previous value minute: 1, client: AAA, location: zoneC.

This is how `having` filter works, not `filter`. The difference between them is, `filter` is applied _before_ aggregation while `having` is applied _after_ aggregation. So, in the above example, with the `not-null` filter, the new row `minute: 1.4, client: AAA, location: null` will be filtered out but other rows can be still consumed by aggregators. The result should be `minute: 1, client: AAA, location: zoneC`.

Reply:
But can you use the filters at indexing time? 

Because if the aggregator at the indexing time tries to index the `minute: 1.4, client: AAA, location: null`. This message discards the lastValue and the client: AAA has `null`location to `minute: 1`, because of the queryGranularity=MINUTE.

Reply:
Oh, you mean ingestion time! Yes, you can do at ingestion time using [FilteredAggregator](http://druid.io/docs/latest/querying/aggregations.html#filtered-aggregator). The filter of `FilteredAggregator` is applied to only the corresponding `Aggregator`. The input data is filtered out before consuming by the `Aggregator`, so that it can read only non-null values.

Reply:
oh!! Really good point. I don't know that I can use the FilteredAggregator at ingestion time! 

Thanks @jihoonson